### DISCUSSION TITLE
workload wasi-experimental - is this where it's supported?  Build tool issue on MacOS M1 - Error MSB4062 : Task could not be loaded from the assembly - WasmAppBuilder.dll.

### DISCUSSION THREAD

**Discussion by Simonl9l** [dotnet/runtime#98538] [Q&A]
It seems `dotnet workloads` is an area of dotnet that is not, to many, well known about it seems. If this is the wrong place, please can you let me know?

I'm running on Apple Silicon (M1/Arm64).

I'm trying to use a project that is dependent on `wasi-experimental` and the [wasi-sdk](https://github.com/WebAssembly/wasi-sdk), and have followed the instruction to instal the MacOs (universal binaries).

It seem that if I follow all this correctly `wasi-experimental` is looking for the `wasi-sdk` in `/Users/{profile}/.wasi-sdk/wasi-sdk-16.0` and, per the project maintainer, putting the `bin`, `lib`, `share` of the v20 sdk in this folder "works on Windows".  I have installed v20 the same way.

I also have the `WASI_SDK_PATH=/Users/{profile}/.wasi-sdk/wasi-sdk-16.0`

If I try and build the project I get the following error:

```
0>Wasi.Sdk.targets(102,3): Error MSB4062 : The "Microsoft.WebAssembly.Build.Tasks.WasmResolveAssemblyDependencies" task could not be loaded from the assembly /usr/local/share/dotnet/packs/Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk/8.0.2/Sdk/../tasks/net8.0/WasmAppBuilder.dll.  
Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
0>------- Finished building project: WasiInterfacePlugin. Succeeded: False. Errors: 1. Warnings: 0
```

Any clues, directions or hits are appreciated as this has me dead in the water.

**Comment by maraf** [comment]
It works with my setup

I have install wasi-experimental workload, downloaded (latest) WASI SDK and put it in `/Users/marekfisera/Bin/wasi-sdk-21.0`, then set `WASI_SDK_PATH`
```
export WASI_SDK_PATH=/Users/marekfisera/Bin/wasi-sdk-21.0
```

Now I can build an app generated from `wasi-console` template with native relink.

Can you do build in your environment and share a binlog (`dotnet build -bl`)?

**Reply by maraf** [reply]
I don't see the use of `WasmResolveAssemblyDependencies` in my build...

**Reply by Simonl9l** [reply]
@maraf thanks of replying please see attached. 
[msbuild.binlog.zip](https://github.com/dotnet/runtime/files/14316769/msbuild.binlog.zip)

```
Installed Workload Id      Manifest Version      Installation Source
--------------------------------------------------------------------
wasi-experimental          8.0.2/8.0.100         SDK 8.0.100        

```

**Reply by maraf** [reply]
The `wasi-experimental` workload is a successor of `Wasi.Sdk` package. Remove the package reference and things should work

**Reply by Simonl9l** [reply]
@maraf ah! Thanks for the spot observation, that was an experiment to fix my issue; but had not realized I'd left it in place. Removing that as you noted solved my issue.

Also if note I actually removed the was-idk install from `/Users/{profile}/.wasi-sdk/wasi-sdk-16.0` and it still worked?

Does this mean that `wasi-experimental` includes its own wasi-sdk?

On the basis the workload is the successor to the sdk, is there any documentation on the `wasi-experimental` workload and what is supports etc ?

Now that it builds, I see that the build (for debug) produces a file structure that has a `bin/net8.0/wasi-wasm/dotnet.wasm` file. I was expecting the `.wasm` file to be named the same as the project?  

Does this need additional project settings to name this as I need?

Thanks again - and apologies I may have created an issue due to my own error, so appreciate you help and observations.

**Reply by maraf** [reply]
No problem, I'm happy to help. The new workload has some rough edges.

> Does this mean that `wasi-experimental` includes its own wasi-sdk?

Not yet. It's something that will most probably change, but at the moment you need to install `wasi-sdk` manually and set `WASI_SDK_PATH` to point to it.

> On the basis the workload is the successor to the sdk, is there any documentation on the wasi-experimental workload and what is supports etc ?

These isn't any official documentation and we don't even have a list of supported MSBuild properties and items yet. You can read through [build targets](https://github.com/dotnet/runtime/blob/release/8.0/src/mono/wasi/build/WasiApp.targets) (please note that not everything mentioned in the comment in the linked file is working for wasi, as it's mostly copied from browser scenario) or ask questions around. We'll get better.

> Now that it builds, I see that the build (for debug) produces a file structure that has a bin/net8.0/wasi-wasm/dotnet.wasm file. I was expecting the .wasm file to be named the same as the project?

The wasi build has two modes. Be default it produces a folder with separate files (dlls), in this mode the wasm file is `dotnet.wasm`, because it contains only .NET runtime. The other mode (`WasmSingleFileBundle=true`) bundles everything in a single wasm file and in this case the name of wasm file matches name of your project.

**Reply by enghch** [reply]
@maraf Do you know if the WasmSingleFileBundle is working on .NET 9? It seemed to work fine for me on .NET 8 but over there, [UnmanagedCallersOnly] does not seem to be implemented yet. But on .NET 9, WasmSingleFileBundle does not roll the files into a single .wasm file.

**Reply by maraf** [reply]
It should be. I'll check that...

**Reply by maraf** [reply]
@enghch Works on my machine (testing with .NET 9 preview6). Are you looking into the `AppBundle` (`bin/Debug/net9.0/wasi-wasm/AppBundle`) directory?

**Reply by enghch** [reply]
Hmm! I'm on .NET 9 preview 7 (windows x64). In the bin\Debug\net9.0\wasi-wasm\AppBundle directory I see dotnet.wasm, node.mjs, run-node.sh, run-wasmtime.sh, and Wasi.runtimeconfig.json. But no other .wasm files.

**Reply by maraf** [reply]
That is the expected state, isn't it?

**Reply by enghch** [reply]
I expected there to be a [my project].wasm file in addition to the dotnet.wasm file. That's how it worked with .NET 8. It would merge my code with the dotnet.wasm code into a single .wasm file. One directory up, I see a [my project].dll file. I think that's the file that should have been converted to wasm. (I just went back to Preview 6 and it appears to have the same behavior.)

The reason I switched from .NET 8 is to 9 is that 8 didn't support the [UnmanagedCallersOnly] attribute. It didn't produce any exports.

**Reply by ReneeGA2020** [reply]
It is supposed to have a dotnet.wasm by default. If you want to generate single project.wasm file, you will have to add `<WasmSingleFileBundle>true</WasmSingleFileBundle> `into you project file. Also you may have to update your wasmtime and wasi-sdk to the latest versions (24 for now)

**Reply by maraf** [reply]
@enghch 

```
    Directory: ..../WasiConsole90/bin/Debug/net9.0/wasi-wasm/AppBundle

drwxr-xr-x marekfisera      staff              02.09.2024 10:16             64 tmp
-rw-r--r-- marekfisera      staff              02.09.2024 10:16            456 node.mjs
-rwxr-xr-x marekfisera      staff              02.09.2024 10:16             82 run-node.sh
-rwxr-xr-x marekfisera      staff              02.09.2024 10:16             35 run-wasmtime.sh
-rw-r--r-- marekfisera      staff              02.09.2024 10:16           1853 WasiConsole90.runtimeconfig.json
-rwxr-xr-x marekfisera      staff              02.09.2024 10:16       33866280 WasiConsole90.wasm
```

This is the expected output for single file. You are right that `dotnet.wasm` should be there, I misread it. 

```diff
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
    <RuntimeIdentifier>wasi-wasm</RuntimeIdentifier>
    <OutputType>Exe</OutputType>
    <PublishTrimmed>true</PublishTrimmed>
+    <WasmSingleFileBundle>true</WasmSingleFileBundle>
  </PropertyGroup>
</Project>
```

This is my csproj, the app is generated from `wasiconsole` template and `WasmSingleFileBundle=true` is added. This behavior is the same as in 8.

**Reply by enghch** [reply]
Thank you! After much experimenting, I finally got this working on .NET 9 Preview 6. I needed to set the OutputType to Exe (I had it as Library). BUT, this still doesn't work for me on Preview 7. On Preview 7, even with the WasmSingleFileBundle=true, it doesn't generate the single wasm file. It just dumps the copy of the dotnet.wasm file and that's all. But I'm unblocked for the moment by using Preview 6!

**Reply by maraf** [reply]
More details here https://github.com/dotnet/runtime/discussions/106350#discussioncomment-10346702.
